File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        boolean[] visited = new boolean[n];

		for (int i = 0; i < n; i++) {
			if (visited[i])
				continue;
			bfs(i, computers, visited);
			answer++;
		}
        
        return answer;
    }
    
    public void bfs(int start, int[][] computers, boolean[] visited){
        Queue<Integer> q = new ArrayDeque<>();
		q.add(start);
		visited[start] = true;

		while (!q.isEmpty()) {
			int poll = q.poll();

			for (int i = 0; i < computers[poll].length; i++) {
				if (poll == i)
					continue; // 자기자신이면
				if (computers[poll][i] == 0)
					continue; // 연결 안되어있으면
				if (visited[i])
					continue; // 이미 방문했으면

				q.offer(i);
				visited[i] = true;
			}
		}
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution(n, computers) {
    let answer = 0;
    const visited = Array(n).fill(0);
    
    const dfs = (index) => {
        
        const computer = computers[index];
        
        for (let i = 0; i < computer.length; i++) {
            if (visited[i]) continue;
            if (computer[i] === 0) continue;
            visited[i] = 1;
            dfs(i);
        }
        
    }
    
    for (let i = 0; i < n; i++) {
        if (visited[i]) continue;
        visited[i] = 1;
        dfs(i);
        answer++;
    }
    
    return answer;
}